### PR TITLE
Fix/unreported bugs and security issues

### PR DESCRIPTION
**Pull request by Sickwoman**
### Summary

This PR addresses **13 previously unreported bugs and security issues** identified during a comprehensive review of both the frontend (React) and backend (FastAPI/Python) codebases.

These fixes improve **security, stability, performance, and maintainability**, with no overlap with existing GitHub issues or PRs.

---

#### Backend

**`tensormap-backend/app/services/deep_learning.py`**

* **#1** Fixed incorrect graph size validation (`sys.getsizeof`) → now uses accurate byte size via `len(raw.encode('utf-8'))`
* **#2** Replaced vertical-stack layout with BFS-based layered layout for better model topology visualization
* **#8** Replaced hardcoded loss selection with extensible `LOSS_MAPPING` dict; invalid types now return HTTP 400
* **#10** Removed deprecated `numpy.int64` usage for NumPy 2.0+ compatibility
* **#14** Added defensive pagination validation at service layer

**`tensormap-backend/app/services/model_run.py`**

* **#3 (CRITICAL / CWE-22)** Fixed path traversal vulnerability in `_helper_generate_file_location`:

  * Rejects `../`
  * Enforces path confinement using `os.path.realpath`
* **#4** Improved Socket.IO reliability:

  * Added `_emit_with_retry` (3 retries + exponential backoff)
  * Emits explicit error event on failure

**`tensormap-backend/app/models/ml.py`**

* **#5** Prevented silent truncation:

  * `parameter` → `VARCHAR(255)`
  * `value` → `VARCHAR(2000)`

**`tensormap-backend/app/services/data_upload.py`**

* **#7** Replaced `Any` with `UploadFile` for proper typing
* **#9** Added fail-fast CSV parsing with HTTP 400 instead of silent failure

**`tensormap-backend/app/schemas/deep_learning.py`**

* **#13** Introduced `KnownOptimizer` and `KnownMetric` enums

  * Enforces validation at request layer
  * Prevents invalid values reaching TensorFlow

**`tensormap-backend/app/routers/deep_learning.py`**

* **#12** Added `request_id` correlation for traceability

  * Included in logs and error responses (HTTP 500)

---

#### Frontend

**`tensormap-frontend/src/containers/Training/Training.jsx`**

* **#6** Fixed circular `useEffect` / stale closure issue using `prevModelListRef`
* **#11** Resolved Socket.IO memory leak:

  * Removed unstable dependencies from effect
  * Eliminated manual `socket.connect()` on disconnect
  * Added proper cleanup of `socketRef` on unmount

---

### Breaking Changes

* `DLModelConfig.optimizer` and `.metric` now strictly accept values from enums:

  * Invalid values will return **HTTP 422**
* `_helper_generate_file_location` now raises `ValueError` for malformed paths
* Database migration required for `ModelConfigs` column size update

---

### Testing

Run backend tests:

```bash
cd tensormap-backend && pytest tests/ -v
```

---

### Notes

* Security vulnerability (#3) is **critical** and should be prioritized for deployment
* Changes are backward-compatible except where explicitly noted

---

### CODE REVIEW

**Comment by ivantha** [comment]
Closing this in line with the TensorMap `gsoc` label convention. Issue #120 reserves the full scope of the GSoC 2026 student's term, and parts of this PR overlap with that scope. Coding for the selected contributor (Kartikey Gupta) begins on 2026-05-25.

Several of the fixes in this PR are independent of the GSoC scope and would be very welcome as a separate, focused PR:

- The path traversal patch in `_helper_generate_file_location` (CWE-22, fix #3)
- NumPy 2.0 compatibility (fix #10)
- CSV fail-fast parsing in `data_upload` (fix #9)
- VARCHAR truncation guard in `models/ml.py` (fix #5)
- `UploadFile` typing in `data_upload` (fix #7)
- BFS-based layered graph layout (fix #2)
- Accurate byte-size validation via `len(raw.encode('utf-8'))` (fix #1)
- `request_id` correlation in `deep_learning` router (fix #12)
- Frontend `useEffect` stale-closure fix and Socket.IO cleanup in `Training.jsx` (fixes #6 and #11)
- Defensive pagination validation (fix #14)

The pieces that overlap with GSoC scope and should be deferred to the student's term:

- The `LOSS_MAPPING` dict (fix #8) starts the loss/layer registry that **Phase 1 (Foundation: layer registry refactor)** is meant to introduce in a unified way.
- The `KnownOptimizer` and `KnownMetric` enums (fix #13) belong with the same registry effort.
- The `_emit_with_retry` Socket.IO retry logic (fix #4) sits inside the **Phase 3 (Training visualization + callbacks)** surface area that is being restructured for structured metrics and real-time charts.

Happy to review a split PR that keeps just the non-overlapping fixes above. The path traversal one is critical and worth landing on its own quickly.
